JIT: yet another OSR stress mode

AndyAyersMS opened the pull request:
Existing OSR stress use existing patchpoint placement logic and alter the
policy settings so that OSR methods are created more eagerly.

This new version of OSR stress alters the patchpoint placement logic, to
add patchpoints to more places. In conjunction with the eager policy stress
above this leads to creation and execution of large numbers of OSR methods.

Any IL offset in the method with an empty stack (and not in a handler) is fair
game for a patchpoint, so this new method randomly adds patchpoints to the
starts of blocks when stack empty (future work may extend this to mid-block
stack empty points).

The new mode is enabled by setting `COMPlus_JitRandomOnStackReplacement` to a
non-zero value; this value represents the likelihood of adding a patchpoint
at a stack-empty block start, and also factors into the random seed. (Recall
this is parsed in hex, so 0x64 == 100 or larger will put a patchpoint at the
start of every stack empty block).

Various values are interesting because once a method reaches a patchpoint
and triggers OSR, the remainder of that method's execution is in the OSR method,
so later patchpoints in the original method may never be reached. So some sort
of random/selective patchpoint approach (in conjunction with varying policy
settings) is needed to ensure that we create an execute as many different OSR
variants as possible.

This PR also includes a couple of fixes exposed by local testing of this new
stress mode:
* The OSR prolog may end up empty, which gcencoder doesn't like. Detect this
and add a `nop` for the prolog.
* If we're importing the `fgEntryBB` during OSR, we don't need to schedule it
for re-importation. This happens if we put a patchpoint at IL offset zero.
* Update the selective dumping policy `COMPlus_JitDumpAtOSROoffset` to only
dump OSR method compilations.

A new test leg is added to `jit-experimental` to run with this mode enabled
with a probability of 21% (0x15) and quick OSR triggers.

AndyAyersMS committed:
JIT: yet another OSR stress mode

Existing OSR stress use existing patchpoint placement logic and alter the
policy settings so that OSR methods are created more eagerly.

This new version of OSR stress alters the patchpoint placement logic, to
add patchpoints to more places. In conjunction with the eager policy stress
above this leads to creation and execution of large numbers of OSR methods.

Any IL offset in the method with an empty stack (and not in a handler) is fair
game for a patchpoint, so this new method randomly adds patchpoints to the
starts of blocks when stack empty (future work may extend this to mid-block
stack empty points).

The new mode is enabled by setting `COMPlus_JitRandomOnStackReplacement` to a
non-zero value; this value represents the likelihood of adding a patchpoint
at a stack-empty block start, and also factors into the random seed. (Recall
this is parsed in hex, so 0x64 == 100 or larger will put a patchpoint at the
start of every stack empty block).

Various values are interesting because once a method reaches a patchpoint
and triggers OSR, the remainder of that method's execution is in the OSR method,
so later patchpoints in the original method may never be reached. So some sort
of random/selective patchpoint approach (in conjunction with varying policy
settings) is needed to ensure that we create an execute as many different OSR
variants as possible.

This PR also includes a couple of fixes exposed by local testing of this new
stress mode:
* The OSR prolog may end up empty, which gcencoder doesn't like. Detect this
and add a `nop` for the prolog.
* If we're importing the `fgEntryBB` during OSR, we don't need to schedule it
for re-importation. This happens if we put a patchpoint at IL offset zero.
* Update the selective dumping policy `COMPlus_JitDumpAtOSROoffset` to only
dump OSR method compilations.

A new test leg is added to `jit-experimental` to run with this mode enabled
with a probability of 21% (0x15) and quick OSR triggers.

AndyAyersMS commented:
cc @dotnet/jit-contrib

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -11776,12 +11782,35 @@ void Compiler::impImportBlockCode(BasicBlock* block)
+            // Reuse the random inliner's random state.
+            // Note m_inlineStrategy is always created, even if we're not inlining.
+            //
+            CLRRandom* const random      = impInlineRoot()->m_inlineStrategy->GetRandom(doRandomOSR);
+            const int        randomValue = (int)random->Next(101);
```
Shouldn't this be `random->Next(100)` for `0x65` to be 100% probability?

AndyAyersMS replied:
Yes.

jakobbotsch reviewed (approved):
LGTM apart from the above.

AndyAyersMS committed:
fix probability calc

AndyAyersMS commented:
/azp run runtime-jit-experimental

AndyAyersMS committed:
remove obsolete assert

AndyAyersMS commented:
Looks like a couple of new failures on arm64 from the new random stress. Have started poking at these.
* `Runtime_1241` -- we're asserting that an osr local can't be a multireg arg; this seems unnecessary as OSR methods are not passed args, they read homed values from the Tier0 frame. Think we can just delete the assert.
* `Runtime_33529` -- jit crash: a recursive tail call gets introduced by inlining, making the original entry block reachable, but OSR didn't import the original entry block. Seems like if we see a tail call that's an inline candidate, we may have to assume it could eventually introduce a tail recursive call, and do speculative importation of the entry path for the root method.

AndyAyersMS committed:
tweak

AndyAyersMS committed:
osr exposed does not apply to simd

AndyAyersMS committed:
only pad zero-sized OSR prolog if we're reporting generic context

AndyAyersMS commented:
Updated with fixes. May still be one or two failures left.

AndyAyersMS commented:
/azp run runtime-jit-experimental

AndyAyersMS committed:
Add ability to have patchpoint at specified offset.

Fix interaction of stress patchpoints and profile instrumentation.
We need to force block-based instrumentation if we might have stress
patchpoints.

AndyAyersMS commented:
/azp run runtime-jit-experimental

AndyAyersMS commented:
@dotnet/jit-contrib PTAL.

There is still one OSR stress failure I'd like to fix, but it can wait for a future PR.

jakobbotsch commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -7542,6 +7536,16 @@ void CodeGen::genFnProlog()
 
 #endif // PROFILING_SUPPORTED
 
+    // For OSR we may have a zero-length prolog. That's not supported
+    // when the method must report a generics context,/ so add a nop if so.
+    //
+    if (compiler->opts.IsOSR() && (GetEmitter()->emitGetPrologOffsetEstimate() == 0) &&
+        (compiler->lvaReportParamTypeArg() || compiler->lvaKeepAliveAndReportThis()))
+    {
+        JITDUMP("OSR: prolog was zero length and has generic context to report: adding nop to pad prolog.\n");
+        instGen(INS_nop);
+    }
```
For my curiosity, why/where is this not supported?

AndyAyersMS replied:
I think it may well be a superfluous assert:
https://github.com/dotnet/runtime/blob/0266f034221188bc7d0156ba76a571c71885c68f/src/coreclr/gcinfo/gcinfoencoder.cpp#L706-L715

The jit only uses `SetPrologSize` if we're reporting a generics context.

jakobbotsch replied:
Ok, perhaps something to relax in another PR.
